src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: LGPL-2.1+
+
+import argparse
+import dataclasses
+import itertools
+import pathlib
+import os
+import shlex
+import subprocess
+import tempfile
+import typing
+
+efi_arches = {
+        # host_arch: efi_arch
+        'x86'    : 'ia32',
+        'x86_64' : 'x64',
+        'arm'    : 'arm',
+        'aarch64': 'aa64',
+        'riscv64': 'riscv64',
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
+    efi_arch = efi_arches[arch]
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
+        except FileNotFoundError:
+            pass
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: str
+    offset: int
+    tmpfile: typing.Optional[typing.IO] = None
+
+    @classmethod
+    def create(cls, name, content, offset):
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            content = tmp.name
+        else:
+            tmp = None
+
+        return cls(name, content, offset, tmpfile=tmp)
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
+        return cls.create(name, content, offset)
+
+
+def parse_args():
+    p = argparse.ArgumentParser(
+        description='Build and sign Unified Kernel Images',
+        usage='''\
+usage: ukify [options…] linux initrd
+       ukify -h | --help
+''')
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
```

Comment:
Currrently `mkinitcpio` and `dracut` implement semantics around `/etc/kernel/cmdline` and `/etc/cmdline.d` which could be implemented in this helper instead. This would remove code duplication. The flag itself could take a singular file, but preferably default to reading `/etc/cmdline.d` is specified.

Reply:
Hmm, but what is the point of those? If anything, I think those should be added to `kernel-install` first.

Reply:
I'm not sure what you are referring to. The initrd generators or the files themself?

Reply:
Sorry for the lack of clarity. I meant that the support for those files when writing a boot loader entry. `kernel-install` looks at `/etc/kernel/cmdline` and for it is the primary tool for dealing with commandline entries. Thus, support for additional configuration files for the command line entry should be added there first.

Reply:
dracut has deprecated `/etc/kernerl/cmdline` for their UKIs. I'm not sure why `kernel-install` is the reference for what is and isn't supported?

Reply:
I'm confused and this was apparently more complicated?

Dracut itself reads from `/etc/cmdline` which is deprecated in favour of `/etc/cmdline.d/*.conf`. But their `kernel-install` hooks do read `/etc/kernel/cmdline`.

I don't understand why all these paths are needed.

Reply:
kernel-install is a systemd thing, and so is /etc/kernel/cmdline.

/etc/cmdline otoh might be something the dracut people came up with?

Reply:
I think so. I do wonder if it's better to standardize these paths for the sake of consistency?

Reply:
`/etc/cmdline` and `/etc/cmdline.d/*.conf` aren't related to kernel-install. dracut uses these paths to allow to store additional kernel cmdline arguments in the initrd itself. When the initrd runs, dracut treats them as if they were part of the real cmdline. (This, obviously, only works if everything accesses the cmdline via dracut-provided functions. Which causes problems these days as some stuff is implemented as units/generators, that only see the cmdline systemd gives them...)